File: Projects/UniverseExplorer/app/src/main/java/a45125/myapplication/TimerBroadcastReceiver.java
package a45125.myapplication;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import static android.content.Context.ALARM_SERVICE;

public abstract class TimerBroadcastReceiver extends BroadcastReceiver {
    private static final int REQUEST_CODE = 101;
    private static int TIMER_ID = 0;
    private final String ACTION_TIMER = "dam_a45125.ACTIONTIMER -"+ TIMER_ID ++;
    private final Context context;
    private final AlarmManager alarmManager;
    private final PendingIntent alarmPendingIntent;

    public TimerBroadcastReceiver ( Context context ) {
        this.context = context;
        alarmManager = (AlarmManager)context.getSystemService(ALARM_SERVICE);
        context.registerReceiver (this , new IntentFilter(ACTION_TIMER));
        alarmPendingIntent = PendingIntent.getBroadcast(context, REQUEST_CODE, new Intent(ACTION_TIMER), 0);
    }

    public void start ( long delaySegs ) {
        long time = System.currentTimeMillis() + delaySegs * 1000;
        alarmManager.setExact(AlarmManager.RTC , time , alarmPendingIntent);
    }

    public void cancel () {
        alarmManager.cancel(alarmPendingIntent);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        onTick();
    }

    public abstract void onTick();

    public void onDestroy() {
        context.unregisterReceiver(this);
    }
}

File: Projects/UniverseExplorer/app/src/main/java/a45125/dam/universeexplorer/TimerBroadcastReceiver.kt
package a45125.dam.universeexplorer

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Context.ALARM_SERVICE
import android.content.Intent
import android.content.IntentFilter


internal abstract class TimerBroadcastReceiver(private val context: Context) : BroadcastReceiver() {
    companion object {
        private const val REQUEST_CODE = 101
        private var TIMER_ID = 0
    }
    private val ACTION_TIMER = "dam_45125.ACTION_TIMER-" + TIMER_ID++
    private val alarmManager: AlarmManager
    private val alarmPendingIntent: PendingIntent

    init {
        alarmManager = context.getSystemService(ALARM_SERVICE) as AlarmManager
        context.registerReceiver(this, IntentFilter(ACTION_TIMER))
        alarmPendingIntent = PendingIntent.getBroadcast(context, REQUEST_CODE, Intent(ACTION_TIMER), 0)
    }


    fun start(delaySegs: Long) {
        val time = System.currentTimeMillis() + delaySegs * 1000
        alarmManager.setExact(AlarmManager.RTC, time, alarmPendingIntent)
    }

    fun cancel() {
        alarmManager.cancel(alarmPendingIntent)
    }

    override fun onReceive(context: Context?, intent: Intent?) {
        onTick()
    }

    abstract fun onTick()
    fun onDestroy() {
        context.unregisterReceiver(this)
    }

}